web/src/components/vis/VolcanoPlot.vue

```diff
@@ -23,106 +46,143 @@ export default {
+      y: -Math.log10(row.pValue),
+    }))) as Ref<TransformedRow[]>;
+
+    const xrange = computed(() => {
+      const max = transformedRows.value
+        .reduce((acc: number, d: TransformedRow) => Math.max(acc, Math.abs(d.x)), 0);
       return [-max, max];
-    },
-    yrange() {
-      const max = this.transformedRows.reduce((acc, d) => Math.max(acc, d.y), 0);
+    }) as Ref<[number, number]>;
+
+    const yrange = computed(() => {
+      const max = transformedRows.value
+        .reduce((acc: number, d: TransformedRow) => Math.max(acc, d.y), 0);
       return [0, max];
-    },
-  },
-  watch: {
-    rows() {
-      this.update();
-    },
-    minFoldChange() {
-      this.update();
-    },
-    minLogP() {
-      this.update();
-    },
-  },
-  mounted() {
-    this.update();
-  },
-  methods: {
-    update() {
-      // Compute the total variance in all the PCs.
-      const svg = select(this.$refs.svg);
-      this.axisPlot(svg);
-
-      svg.select('.plot').selectAll('circle')
-        .data(this.transformedRows)
-        .join((enter) => {
-          const r = enter.append('circle');
-          r.append('title');
-          return r;
-        })
-        .attr('r', (d) => (Math.abs(d.x) >= this.minFoldChange && d.y >= this.minLogP ? this.radius * 2 : this.radius))
-        .attr('opacity', (d) => (Math.abs(d.x) >= this.minFoldChange && d.y >= this.minLogP ? 1 : 0.5))
-        .attr('cx', (d) => this.scaleX(d.x))
-        .attr('cy', (d) => this.scaleY(d.y))
-        .style('fill', (d) => d.color)
-        .select('title')
-        .text((d) => `${d.name}: ${d.log2FoldChange} x ${d.pValue}`);
-      svg.select('.plot').selectAll('line.x-threshold')
-        .data([-1, 1])
-        .join('line')
-        .attr('class', 'x-threshold')
-        .style('stroke', 'black')
-        .style('stroke-width', 0.5)
-        .attr('x1', (d) => this.scaleX(d * this.minFoldChange))
-        .attr('y1', 0)
-        .attr('x2', (d) => this.scaleX(d * this.minFoldChange))
-        .attr('y2', this.height - this.margin.top - this.margin.bottom);
-      svg.select('.plot').selectAll('line.y-threshold')
-        .data([1])
-        .join('line')
-        .attr('class', 'y-threshold')
-        .style('stroke', 'black')
-        .style('stroke-width', 0.5)
-        .attr('x1', 0)
-        .attr('y1', () => this.scaleY(this.minLogP))
-        .attr('x2', this.width - this.margin.left - this.margin.right)
-        .attr('y2', () => this.scaleY(this.minLogP));
-    },
+    }) as Ref<[number, number]>;
+
+    const scaleX = computed(() => scaleLinear()
+      .domain(xrange.value)
+      .range([0, dwidth.value]));
+
+    const scaleY = computed(() => scaleLinear()
+      .domain(yrange.value)
+      .range([dheight.value, 0]));
+
+    const axisX = computed(() => axisBottom(scaleX.value));
+    const axisY = computed(() => axisLeft(scaleY.value));
+
+    onMounted(() => {
+      watchEffect(() => {
+        // Recalculate width/height
```

Comment:
Even more than with #604 I think it'd be good to hoist dependencies to the top of this function. 

D3 style makes it hard (for me) to figure out where looping is happening, and reading from a reactive variable incurs a cost that can be amplified to noticeable slowness in a loop, so you might even get better performance by taking the read penalty once at the top.

Reply:
Agreed, I moved the dependencies to the top of the `watchEffect` function.